agent/src/agent_win.cc

```diff
@@ -11,87 +11,383 @@
+  if (success) {
+    return CallHandler();
+  }
+
+  // If success os false, there are two possibilities:
+  //
+  // 1/ The last error is ERROR_MORE_DATA(234).  This means there are more
+  //     bytes to read before the request message is complete.  Resize the
+  //     buffer and adjust the cursor.  The caller will queue up another read
+  //     and wait.
+  // 2/ Some error occured.  In this case return the error.
+
+  DWORD err = GetLastError();
+  if (err == ERROR_MORE_DATA) {
+    read_size_ = kBufferSize;
+    buffer_.resize(buffer_.size() + read_size_);
+    cursor_ = buffer_.data() + buffer_.size() - read_size_;
+  }
+
+  return err;
+}
+
+DWORD AgentWin::Connection::CallHandler() {
+  DWORD err = ERROR_SUCCESS;
+
+  ChromeToAgent request;
+  if (request.ParseFromArray(buffer_.data(), final_size_)) {
+    if (request.has_request()) {
+      auto event = std::make_unique<ContentAnalysisEventWin>(
+          handle_, browser_info_, std::move(*request.mutable_request()));
+      if (event->Init() == ERROR_SUCCESS) {
+        handler_->OnAnalysisRequested(std::move(event));
+      } else {
+        err = ERROR_INTERNAL_ERROR;
+      }
+    } else if (request.has_ack()) {
+      handler_->OnResponseAcknowledged(request.ack());
+    } else {
+      // Malformed request.
+      err = ERROR_INVALID_DATA;
+    }
+  }
+
+  return err;
+}
+
+DWORD AgentWin::Connection::BuildBrowserInfo() {
+  if (!GetNamedPipeClientProcessId(handle_, &browser_info_.pid)) {
+    return GetLastError();
+  }
+
+  HANDLE hProc = OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION, FALSE,
+      browser_info_.pid);
+  if (hProc == nullptr) {
+    return GetLastError();
+  }
+  
+  DWORD err = ERROR_SUCCESS;
+  char path[MAX_PATH];
+  DWORD size = sizeof(path);
+  DWORD length = QueryFullProcessImageNameA(hProc, 0, path, &size);
+  if (length == 0) {
+    err = GetLastError();
+  }
+
+  CloseHandle(hProc);
+
+  browser_info_.binary_path = path;
+  return err;
+}
+
+AgentWin::AgentWin(
+    Config config,
+    std::unique_ptr<AgentEventHandler> event_handler)
+  : AgentBase(std::move(config), std::move(event_handler)) {
+  if (handler() == nullptr)
+    return;
+
   std::string pipename =
-      internal::GetPipeName(configuration().name, configuration().user_specific);
+      internal::GetPipeName(configuration().name,
+                            configuration().user_specific);
   if (pipename.empty())
     return;
 
   pipename_ = pipename;
-  CreatePipe(&hPipe_);
+
+  connections_.reserve(kNumPipeInstances);
+  for (int i = 0; i < kNumPipeInstances; ++i) {
+    connections_.emplace_back(
+        std::make_unique<Connection>(pipename_, handler(), i == 0));
+    if (!connections_.back()->IsValid()) {
+      Shutdown();
+      break;
+    }
+  }
 }
 
 AgentWin::~AgentWin() {
   Shutdown();
 }
 
-std::unique_ptr<Session> AgentWin::GetNextSession() {
-  std::unique_ptr<SessionWin> session;
+void AgentWin::HandleEvents() {
+  std::vector<HANDLE> wait_handles;
+  for (;;) {
```

Comment:
C++ noob question:  Is this shorthand for lopping over `wait_handles`?

Reply:
This is a "forever loop".  Changed to `while (true)` instead.